src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -0,0 +1,300 @@
+
+namespace System.Text.Json.Serialization
+{
+    internal abstract class JsonPropertyInfo
+    {
+        protected static readonly JsonDateConverterAttribute s_dateConverterAttibute = new JsonDateConverterAttribute();
+        protected static readonly JsonEnumConverterAttribute s_enumConverterAttibute = new JsonEnumConverterAttribute();
+        protected static readonly JsonArrayConverterAttribute s_arrayConverterAttibute = new JsonArrayConverterAttribute();
+
+        public PropertyNamePolicyAttribute NameConverter;
+        public bool IncludeNullValues;
+        public byte[] Key = null;
+
+        public JsonPropertyInfo(Type classType, Type propertyType, PropertyInfo propertyInfo, Type elementType, JsonConverterSettings options)
+        {
+            ClassType = classType;
+            PropertyType = propertyType;
+            PropertyInfo = propertyInfo;
+            ElementType = elementType;
+            Options = options;
+        }
+
+        public EnumerableConverterAttribute EnumerableConverter { get; private set; }
+
+        public JsonConverterSettings Options { get; private set; }
+
+        public PropertyInfo PropertyInfo { get; private set; }
+
+        public Type PropertyType { get; private set; }
+
+        public Type ClassType { get; private set; }
+
+        public Type ElementType { get; private set; }
+
+        public abstract Type TempListType { get; }
+
+        public abstract object GetValueAsObject(object obj);
+        public abstract void SetValueAsObject(object obj, object value);
+
+        public abstract PropertyValueConverterAttribute GetValueConverter();
+
+        public virtual void GetPolicies()
+        {
+            {
+                PropertyNamePolicyAttribute attr = GetPolicy<PropertyNamePolicyAttribute>();
+                if (attr != null)
+                {
+                    NameConverter = attr;
+                }
+            }
+
+            {
+                IncludeNullValues = GetPropertyClassAssemblyPolicy(attr => attr.IncludeNullValues).GetValueOrDefault();
+            }
+
+            if (ElementType != null)
+            {
+                EnumerableConverter = GetEnumerableConverter();
+            }
+        }
+
+        protected TAttribute GetPolicy<TAttribute>() where TAttribute : Attribute
+        {
+            TAttribute attr = null;
+            if (PropertyInfo != null)
+            {
+                // Use Property first
+                attr = Options.GetAttributes<TAttribute>(PropertyInfo).FirstOrDefault();
+            }
+
+            if (attr == null)
+            {
+                // Then class type
+                attr = Options.GetAttributes<TAttribute>(ClassType, inherit: true).FirstOrDefault();
+
+                if (attr == null)
+                {
+                    // Then declaring assembly
+                    attr = Options.GetAttributes<TAttribute>(ClassType.Assembly).FirstOrDefault();
+                }
+            }
```

Comment:
This can be simplified to:

```csharp
TAttribute attr = PropertyInfo != null
    ? Options.GetAttributes<TAttribute>(PropertyInfo).FirstOrDefault()
    : null;
return attr
    ?? Options.GetAttributes<TAttribute>(ClassType, inherit: true).FirstOrDefault();
    ?? Options.GetAttributes<TAttribute>(ClassType.Assembly).FirstOrDefault();
```